### ISSUE TITLE
Newly created configurations are sometimes assigned to occupied port

### ISSUE THREAD

**Issue by vpavic** [openid-certification/oidctest#53]
I've also noticed that newly created configurations are sometimes assigned to a port that is already occupied.

More specifically, this appears to happen when creating a configuration shortly after some other was deleted - in that scenario, the new configuration get assigned to a port however the actual configuration settings are the ones of a previously deleted configuration, as if the deletion operation only unassigned the port but did not purge the configuration itself.

Initially the problem was observed on instance running at https://op.certification.openid.net:60000/ however afterwards we were able to reproduce it using local Docker compose based instance.

Let me know if any further info is needed.

Thanks.

**Comment by Spomky** [comment]
Hi,

I encounter the same issue when I ty to add my application at https://op.certification.openid.net:60000/

**Comment by zandbelt** [comment]
It seems that the server configuration got out of sync.

@rohe: I've now killed the running processes that had no entry in the `assigned_ports.json` and `entities` directory. That should solve the immediate problem as those ports are no longer occupied by "zombie" entities. As you suggested we should look into the root cause: it may have to do with input checking on tags that contain special characters.

@vpavic @Spomky you should be able to register a new instance now.

**Comment by Spomky** [comment]
@zandbelt It works fine now. Thanks.

**Comment by vpavic** [comment]
@zandbelt @rohe Thanks for addressing the problem with https://op.certification.openid.net:60000/.

My colleague and I have spent some time trying to reproduce the problem in a reproducible manner using local Docker compose based instance, and we've run into conclusion that there's something wrong with the process of deleting configurations.

Note that this is running a fresh build from the current `master` branch. Steps to reproduce the problem are:

1) create a new configuration
2) delete newly created configuration
3) create another configuration with different tag from the one used in step 1
4) newly created configuration will point to configuration that created in step 1 (and deleted in step 2)

The problem does not occur if the tag name of configuration created in step 3 is the same as in step 1.

**Comment by zandbelt** [comment]
Excellent work: being able to reproduce this will greatly help us addressing the root cause. Many thanks!

**Comment by zandbelt** [comment]
@rohe: looks like the issue is that a "delete" doesn't first kill the process for a running instance. The code seems to confirm that: https://github.com/openid-certification/oidctest/blob/master/src/oidctest/tt/action.py#L306
since that method is only called with `pid=0`.

So it's probably a matter of aligning the code in "delete" method with the code in the "stop" method (since the latter does work), agree?

**Comment by rohe** [comment]
> 3 okt. 2017 kl. 16:49 skrev Hans Zandbelt <notifications@github.com>:
> 
> @rohe <https://github.com/rohe>: looks like the issue is that a "delete" doesn't first kill the process for a running instance. The code seems to confirm that: https://github.com/openid-certification/oidctest/blob/master/src/oidctest/tt/action.py#L306 <https://github.com/openid-certification/oidctest/blob/master/src/oidctest/tt/action.py#L306>
> since that method is only called with pid=0.
> 
OK
> So it's probably a matter of aligning the code in "delete" method with the code in the "stop" method (since the latter does work), agree?
> 
> 
Sounds reasonable !

— Roland

**Comment by zandbelt** [comment]
@vpavic: the proposed fix is in the master branch now; I've verified that it kills the running process for an instance before deleting its config on disk; if you could confirm that this solves your issue we can make a plan to deploy it with the next release on the production server.

**Comment by vpavic** [comment]
Thanks for the very quick resolution! We'll do another fresh build tomorrow to repeat our tests and provide you the feedback.

**Comment by vpavic** [comment]
@zandbelt We just did another build and test run, and we can confirm that the problem is fixed.

However we've encountered one problem while running the build:

```
Step 12 : RUN cd pyoidc && git checkout tags/v0.12.0 -b temp-patch
 ---> Running in 874e01edc098
fatal: Cannot update paths and switch to branch 'temp-patch' at the same time.
Did you intend to checkout 'tags/v0.12.0' which can not be resolved as commit?
ERROR: Service 'op_test' failed to build: The command '/bin/sh -c cd pyoidc && git checkout tags/v0.12.0 -b temp-patch' returned a non-zero code: 128
```

The workaround was to comment out [these lines in op_test Dockerfile](https://github.com/openid-certification/oidctest/blob/master/docker/op_test/Dockerfile#L18-L22).

**Comment by zandbelt** [comment]
I believe that is due to the Docker cache. The pyoidc master branch has been updated in the mean time and Docker cannot know. You'll need to build the docker containers from scratch.

**Comment by vpavic** [comment]
> You'll need to build the docker containers from scratch.

We did that - with each fresh build we removed both old containers and images.

**Comment by zandbelt** [comment]
well, the tag exists in pyoidc, the CI tests are OK and running the offending commands manually also works fine, so IMHO cannot be anything else than a local caching issue

**Comment by vpavic** [comment]
I've tried it with a fresh clone and build on a different machine and it does work.

So the issue was indeed due to some local caching - thanks for the support and sorry for the noise.

**Comment by vpavic** [comment]
Hi @zandbelt, I assume this ticket was reopened due to the fix not being deployed to production instance at https://op.certification.openid.net:60000/?

I was just about to follow up here as I'm having problems adding new configurations for my issuer. New configurations are constantly getting linked to zombie instances.

**Comment by zandbelt** [comment]
yes, I'm sorry; I'll aim for releasing that fix soon as part of version 2.0.8; probably next wednesday

**Comment by vpavic** [comment]
Thanks for the quick feedback.
